aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,349 @@
+    const EZERO_TARGET_USAGE: u64 = 3;
+    const ETARGET_USAGE_TOO_BIG: u64 = 4;
+    const EINVALID_CURVE_LENGTH: u64 = 5;
+    const EINVALID_CURVE_START_POINT: u64 = 6;
+    const EINVALID_CURVE_END_POINT: u64 = 7;
+    const EINVALID_MONOTONICALLY_NON_DECREASING_CURVE: u64 = 8;
+
+    const BASIS_POINT_DENOMINATION: u64 = 10000;
+
+    /// This updates at reconfig and guarantees not to change elsewhere, safe
+    /// for gas calculation.
+    ///
+    /// Specifically, it is updated by executing a reconfig transaction based
+    /// on the usage at the begining of the current epoch. The gas schedule
+    /// derived from these parameter will be for gas calculation of the entire
+    /// next epoch.
+    /// -- The data is one epoch older than ideal, but VM doesn't need to worry
+    /// about reloading gas parameters after the first txn of an epoch.
+    struct StorageGas has key {
+        per_item_read: u64,
+        per_item_create: u64,
+        per_item_write: u64,
+        per_byte_read: u64,
+        per_byte_create: u64,
+        per_byte_write: u64,
+    }
+
+    // x and y are basis points.
+    struct Point has copy, drop, store {
+        x: u64,
+        y: u64
+    }
+
+    /// P(x) = min_price + (base ^ (utilization / target_usage) - 1) / (base - 1) * (max_price - min_price)
+    struct UsageGasConfig has copy, drop, store {
+        target_usage: u64,
+        read_curve: GasCurve,
+        create_curve: GasCurve,
+        write_curve: GasCurve,
+    }
+
+    /// The curve assumes there are two points (0, 0) and (10000, 10000) on both ends. Moreover, points must also
+    /// satisfy the following rules:
+    /// 1. the x values must be strictly increasing and between (0, 10000);
+    /// 2. the y values must be non-decreasing and between (0, 10000);
+    /// So the curve will be identified as point (0, 0) and (10000, 10000) interpolated with the points. The y value
+    /// between two points will be calculated by neighboring points as if there is a linear line connecting these two
+    /// points.
+    struct GasCurve has copy, drop, store {
+        min_gas: u64,
+        max_gas: u64,
+        points: vector<Point>,
+    }
+
+    // Provide a default exponential curve with the base to be 32, which means:
+    // When DB is at 50% target usage,the price increases roughly 15% of (max_price - min_price).
+    // Detailed data points:
+    //   10% -> 1%
+    //   20% -> 3%
+    //   30% -> 6%
+    //   40% -> 10%
+    //   50% -> 15%
+    //   60% -> 23%
+    //   70% -> 33%
+    //   80% -> 48%
+    //   90% -> 70%
+    //   95% -> 84%
+    //   99% -> 96%
+    public fun base_32_exponential_curve(min_gas: u64, max_gas: u64): GasCurve {
+        new_gas_curve(min_gas, max_gas,
+            vector[
+                Point {x: 1000, y: 100},
+                Point {x: 2000, y: 300},
+                Point {x: 3000, y: 600},
+                Point {x: 4000, y: 1000},
+                Point {x: 5000, y: 1500},
```

Comment:
@msmouse when we discussed previously, you said something like "we want the the cost to be 1.15x of the base cost". However here it's base + 0.15x the max price hike. Just to double check: which one is correct?

Reply:
> 0.15x the max price hike
^ is what I meant, sorry